pkg/relay/ethereum/tx.go

```diff
@@ -457,3 +413,242 @@ func (c *Chain) getRevertReasonFromEstimateGas(err error) (string, []byte, error
+			return nil, err
+		}
+
+		if rawTxData, err := tx.MarshalBinary(); err != nil {
+			logger.Error("failed to encode tx", err)
+		} else {
+			logger.Logger = logger.With(logAttrRawTxData, hex.EncodeToString(rawTxData))
+		}
+
+		txGasLimit, isBreak, err := estimateGas(ctx, c, tx, false, logger)
+		if err != nil {
+			return nil, err
+		}
+		if isBreak {
+			return nil, fmt.Errorf("isBraek should not be false")
+		}
+		opts.GasLimit = txGasLimit
+	}
+
+	opts.NoSend = false
+	tx, err := c.SendTx(opts, *iter.Current(), iter.skipUpdateClientCommitment)
+	if err != nil {
+		logger.Error("failed to send msg", err)
+		return nil, err
+	}
+	iter.Next()
+	return tx, nil
+}
+
+func (iter *CallIter) sendMultiTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+	from := iter.Cursor()
+	if (iter.End()) {
+		return nil, nil
+	}
+
+	logger := c.GetChainLogger()
+
+	var (
+		saveTxGasLimit uint64
+		saveRawTxData string
+	)
+	calls := make([]multicall3.Multicall3Call, 0, len(iter.msgs))
+
+	opts, err := c.TxOpts(ctx, true);
+	if err != nil {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for !iter.End() && saveTxGasLimit < c.Config().MaxGasLimit {
+		logger := &log.RelayLogger{Logger: logger.With(
+			logAttrMsgIndexFrom, from,
+			logAttrMsgIndexTo, iter.Cursor(),
+			logAttrMsgType, fmt.Sprintf("%T", *iter.Current()),
+		)}
+
+		opts.GasLimit = math.MaxUint64
+		opts.NoSend = true
+		tx, err := c.SendTx(opts, *iter.Current(), iter.skipUpdateClientCommitment)
+		if err != nil {
+			if len(calls) > 0 {
+				break
+			}
+			logger.Error("failed to build tx for gas estimation", err)
+			return nil, err
+		}
+
+		to := tx.To(); if to == nil {
+			if len(calls) > 0 {
+				break
+			}
+			err2 := fmt.Errorf("no target address")
+			logger.Error("failed to construct Multicall3Call", err2)
+			return nil, err2
+		}
+
+		newCalls := append(calls, multicall3.Multicall3Call{
+			Target: *tx.To(),
+			CallData: tx.Data(),
+		})
+		multiTx, err := c.multicall3.Aggregate(opts, newCalls)
+		if err != nil {
+			if len(calls) > 0 {
+				break
+			}
+			logger.Error("failed to call Multicall3.Aggregate", err)
+			return nil, err
+		}
+
+		if rawTxData, err := multiTx.MarshalBinary(); err != nil {
+			logger.Error("failed to encode multiTx", err)
+		} else {
+			saveRawTxData = hex.EncodeToString(rawTxData)
+			logger.Logger = logger.With(logAttrRawTxData, saveRawTxData)
+		}
+
+		txGasLimit, isBreak, err := estimateGas(ctx, c, multiTx, 0 < len(calls), logger)
+		if err != nil {
+			return nil, err
+		}
+		if isBreak { // tx is fail or gas overs limit and 0 < len(calls)
+			break // send txs with last calls
+		}
+
+		// this calls is ok. save it and try to include next call
+		saveTxGasLimit = txGasLimit
+		calls = newCalls
+		iter.Next()
+	}
+	// now len(calls) > 0
+
+	logger = &log.RelayLogger{Logger: logger.With(
+		logAttrMsgIndexFrom, from,
+		logAttrMsgIndexTo, iter.Cursor(),
+		logAttrRawTxData, saveRawTxData,
```

Comment:
This `logger` is used for logging information related to the result of `eth_sendTransaction`.
On the other hand, `saveRawTxData` stores tx information used to call `eth_estimateGas`.
Therefore I think it's not necessary to add `saveRawTxData` to the logger context.

I think the part below can be moved into `func estimateGas`.
```
if rawTxData, err := tx.MarshalBinary(); err != nil {
	logger.Error("failed to encode tx", err)
} else {
	logger.Logger = logger.With(logAttrRawTxData, hex.EncodeToString(rawTxData))
}
```

Reply:
fixed